src/main/java/io/tiledb/spark/TileDBDataReaderPartitionScan.java

```diff
@@ -224,41 +413,76 @@ public boolean next() {
   public ColumnarBatch get() {
     metricsUpdater.startTimer(queryGetTimerName);
     try {
-      int colIdx = 0;
-      int nRows = 0;
-      // This is a special case for COUNT() on the table where no columns are materialized
-      if (sparkSchema.fields().length == 0) {
-        // TODO: materialize the first dimension and count the result set size
-        try (Dimension dim = domain.getDimension(0)) {
-          if (dim.isVar()) {
-            nRows =
-                getVarLengthAttributeColumn(
-                    dim.getName(), dim.getType(), dim.isVar(), dim.getCellValNum(), 0);
-          } else {
-            nRows = getScalarValueColumn(dim.getName(), dim.getType(), 0);
-          }
-        }
-      } else {
-        // loop over all Spark attributes (DataFrame columns) and copy the query result set
-        for (StructField field : sparkSchema.fields()) {
-          nRows = getColumnBatch(field, colIdx);
-          colIdx++;
-        }
+      int nRows = (int) currentNumRecords;
+      if (resultBatch == null) {
+        ColumnVector[] colVecs = new ColumnVector[valueValueVectors.size()];
+        for (int i = 0; i < valueValueVectors.size(); i++) {
+          String name = fieldNames.get(i);
+          TypeInfo typeInfo = getTypeInfo(name);
+          boolean isDateType = typeInfo.multiplier != 1 || typeInfo.moreThanDay;
+
+          // if nullable
+          if (typeInfo.isNullable) {
+            // If the attribute is nullable we need to set the validity buffer from the main value
+            // vector in bitmap fashion.
+            // TileDB handles the bitmap as a bytemap, thus the following conversion.
+            ArrowBuf arrowBufValidity = valueValueVectors.get(i).getValidityBuffer();
+            for (int j = 0; j < nRows; j++) {
+              if (validityValueVectors.get(i).getDataBuffer().getByte(j) == (byte) 0) {
```

Comment:
This seems like it's going to be really slow for large data - does this get optimized? Can we help it by pulling `validityValueVectors.get(i).getDataBuffer()` outside the loop into a constant reference?